rust-plugins/sass/scripts/setup.mjs

```diff
@@ -0,0 +1,93 @@
+import * as fs from 'node:fs/promises';
+import * as p from 'node:path';
+import { spawnSync } from 'node:child_process';
+import fetch from 'node-fetch';
+import extractZip from 'extract-zip';
+import { extract as extractTar } from 'tar';
+
+const OS = (() => {
+  const platform = process.platform;
+  switch (platform) {
+    case 'linux':
+      return 'linux';
+    case 'darwin':
+      return 'macos';
+    case 'win32':
+      return 'windows';
+    default:
+      throw Error(`Platform ${platform} is not supported.`);
+  }
+})();
+
+const ARCH = (() => {
+  const arch = process.arch;
+  switch (arch) {
+    case 'ia32':
+      return 'ia32';
+    case 'x86':
+      return 'ia32';
+    case 'x64':
+      return 'x64';
+    case 'arm64':
+      return 'arm64';
+    default:
+      throw Error(`Architecure ${arch} is not supported.`);
+  }
+})();
+
+const ARCHIVE_EXTENSION = OS === 'windows' ? '.zip' : '.tar.gz';
+
+async function cleanDir(dir) {
+  await fs.mkdir(p.dirname(dir), { recursive: true });
+  try {
+    await fs.rm(dir, { recursive: true });
+  } catch (_) {
+    // If dir doesn't exist yet, that's fine.
+  }
+}
+
+async function download(url) {
+  console.log(`Downloading "${url}".`);
+  const response = await fetch(url, { redirect: 'follow' });
+  if (!response.ok) {
+    throw Error(`Failed to download ${url}: ${response.statusText}`);
+  }
+  return Buffer.from(await response.arrayBuffer());
+}
+
+async function getDartSassEmbedded(outPath, dirName) {
+  const version = JSON.parse(
+    await fs.readFile(p.join('./ext/sass/package.json'), 'utf8')
+  ).dependencies['sass-embedded'];
+
+  const releaseAsset = await download(
+    `https://github.com/sass/dart-sass-embedded/releases/download/${version}/sass_embedded-${version}-${OS}-${ARCH}${ARCHIVE_EXTENSION}`
+  );
+
+  console.log(`Unzipping dart-sass-embedded release asset to ${outPath}.`);
+  await cleanDir(p.join(outPath, dirName));
+  const zippedAssetPath = `${outPath}/${dirName}${ARCHIVE_EXTENSION}`;
+  await fs.writeFile(zippedAssetPath, Buffer.from(releaseAsset));
+  if (OS === 'windows') {
+    await extractZip(zippedAssetPath, {
+      dir: p.join(process.cwd(), outPath),
+    });
+  } else {
+    extractTar({
+      file: zippedAssetPath,
+      cwd: outPath,
+      sync: true,
+    });
+  }
+  await fs.unlink(zippedAssetPath);
+  await fs.rename(p.join(outPath, 'sass_embedded'), p.join(outPath, dirName));
+}
+
+(async () => {
+  try {
+    await getDartSassEmbedded('./ext/sass', 'sass-embedded');
+  } catch (error) {
+    console.error(error);
+    process.exitCode = 1;
+  }
+})();
```

Comment:
This file will run on the user's machine

1. Will it block Install because of network problems (such as China)

2. It depends on `node`. Is there any compatibility problem without compile (Can a minimum version be given?)

Reply:
I think we can provide a internal binary as downgrade when download fail

Reply:
There is no good way to deal with the first point, unless we provide a downgrade file as mentioned in 5, but then we need to pull a copy of each OS file, otherwise we can't get compatibility with each system. But then it will increase the size of the plugin package.

I think the second point is to restrict the node version >= 16 in package.json, the script itself does not need to be compatible.

I wonder if there is a better way to solve this problem?

Reply:
> This file will run on the user's machine
> 
> 1. Will it block Install because of network problems (such as China)
> 2. It depends on `node`. Is there any compatibility problem without compile (Can a minimum version be given?)

There is no good way to deal with the first point, unless we provide a downgrade file as mentioned @wre232114 , but then we need to pull a copy of each OS file, otherwise we can't get compatibility with each system. But then it will increase the size of the plugin package.

I think the second point is to restrict the node version >= 16 in package.json, the script itself does not need to be compatible.

I wonder if there is a better way to solve this problem?

Reply:
1. I prefer to place executables in the packages of each system under the npm folder, so that executables are only downloaded at release time (ci), and the executables are distributed by the image server, reducing network problems.

Of course, this is just a suggestion, and you don't have to follow it.